### PR TITLE
port deseq2 to Python

### PR DESCRIPTION
**Pull request by EpigeneMax**
This PR integrates our Python port of the R package DESeq2 into a new module of `inmoose`.

Note that not all features of the original R package have been ported yet.

### CODE REVIEW

File: tests/deseq2/test_dispersions.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        # test fit alternative
+        dds = makeExampleDESeqDataSet()
+        dds = dds.estimateSizeFactors()
+        #ddsLocal = dds.copy().estimateDispersions(fitType="local")
+        ddsMean = dds.copy().estimateDispersions(fitType="mean")
+        ddsMed = estimateDispersionsGeneEst(dds.copy())
+        useForMedian = ddsMed.var["dispGeneEst"] > 1e-7
+        medianDisp = np.nanmedian(ddsMed.var["dispGeneEst"][useForMedian])
+        ddsMed.setDispFunction(lambda x: medianDisp)
+        ddsMed = estimateDispersionsMAP(ddsMed)
```
Comment (soleneweill) [inline tests/deseq2/test_dispersions.py:147]: Why is there not assertion on this object if we want to try it?

File: tests/deseq2/test_collapse.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import unittest
+import numpy as np
+
+from inmoose.deseq2 import makeExampleDESeqDataSet, collapseReplicates
+
+class Test(unittest.TestCase):
+
+    def test_collapse(self):
```
Comment (soleneweill) [inline tests/deseq2/test_collapse.py:8]: Can you add a small description as you did for the other functions please

File: tests/deseq2/test_DESeq.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
+import unittest
+import patsy
+
+from inmoose.utils import Factor
+from inmoose.deseq2 import DESeq, makeExampleDESeqDataSet
+
+class Test(unittest.TestCase):
+
+    def test_DESeq(self):
```
Comment (soleneweill) [inline tests/deseq2/test_DESeq.py:9]: Shouldn't we also test that deseq give the good results?

**Review by soleneweill — COMMENTED** [review]
Impressive work. I read the tests as you explained. I have some comments. Mostly to understand. 
As discussed it would be nice to add plots (volcano plot and MA plot at least)

**Comment by soleneweill** [comment]
There is no documentation for deseq2 on the readme :)

File: tests/deseq2/test_linear_mu.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+        dds = dds.estimateSizeFactors()
+        dds1 = DESeqDataSet(dds.copy())
+        dds1 = estimateDispersionsGeneEst(dds1, linearMu=False)
+        dds2 = estimateDispersionsGeneEst(dds.copy(), linearMu=True)
+        mu1 = dds1.layers["mu"]
+        mu2 = dds2.layers["mu"]
+        cors = np.diag(np.corrcoef(mu1, mu2)[:mu1.shape[0], mu1.shape[1]:])
+        self.assertTrue(np.all(cors > 1 - 1e-6))
+
+        dds2 = estimateDispersionsFit(dds2, fitType="mean")
```
[the remark this answers is not in the document]
Comment (EpigeneMax) [reply tests/deseq2/test_linear_mu.py:22]: the idea here is to control that the computation runs smoothly (without error).

File: tests/deseq2/test_weights.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        self.assertEqual(dds2.results().lfcSE[0], dds3.results().lfcSE[0])
+        self.assertEqual(dds2.var["deviance"][0], dds3.var["deviance"][0])
+
+    def test_weights_downweight_outliers(self):
+        """test that weights downweight outlier in dispersion estimation"""
+        dds = makeExampleDESeqDataSet(n=10, seed=42)
+        dds.counts()[0,0] = 100
+        dds.sizeFactors = np.ones(dds.n_obs)
+        dds = dds.estimateDispersions()
+        dds2 = dds.copy()
```
[the remark this answers is not in the document]
Comment (EpigeneMax) [reply tests/deseq2/test_weights.py:87]: first line of `dds` contains an outlier.
`dds` is unweighted, `dds2` is weighted and in `dds3` we remove the outlier.
The idea of the test is to compare the consistency between the dispersion estimates of those three versions.

File: tests/deseq2/test_weights.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        # MAP estimates won't be equal because of different dispersion prior widths
+        self.assertGreater(dds.var["dispMAP"][0], dds2.var["dispMAP"][0])
+
+    def test_weights_warning(self):
+        """test that weights failing check gives warning, passes them through"""
+        dds = makeExampleDESeqDataSet(n=10, seed=42)
+        w = np.ones(dds.shape)
+        w[0:6,0] = 0
+        dds.layers["weights"] = w
+        with self.assertWarnsRegex(UserWarning, expected_regex="for 1 genes, the weights as supplied won't allow parameter estimation"):
```
[the remark this answers is not in the document]
Comment (EpigeneMax) [reply tests/deseq2/test_weights.py:107]: The test just checks part of the warning. The [complete warning message](https://github.com/epigenelabs/inmoose/blob/deseq2/inmoose/deseq2/weights.py#L92) is more explicit.

File: tests/deseq2/test_results.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+            dds.results(addMLE=True)
+        with self.assertRaisesRegex(ValueError, expected_regex="tests of log fold change above or below a theshold must be Wald tests"):
+            dds.results(lfcThreshold=1, test="LRT")
+
+        self.assertTrue(np.all(
+            dds.results(test="LRT", contrast=["group", "1", "2"]).log2FoldChange == - dds.results(test="LRT", contrast=["group", "2", "1"]).log2FoldChange))
+
+
+    @unittest.skip("not sure what to test")
+    def test_results_basics(self):
```
[the remark this answers is not in the document]
Comment (EpigeneMax) [reply tests/deseq2/test_results.py:148]: updated

File: tests/deseq2/test_dispersions.py (hunk 1 above)
Comment (EpigeneMax) [reply tests/deseq2/test_dispersions.py:147]: The intent here is rather to test that no error is raised, that `estimateDispersionsMAP` has enough info to run

File: tests/deseq2/test_collapse.py (hunk 2 above)
Comment (EpigeneMax) [reply tests/deseq2/test_collapse.py:8]: 👍

File: tests/deseq2/test_DESeq.py (hunk 3 above)
Comment (EpigeneMax) [reply tests/deseq2/test_DESeq.py:9]: Agreed. So far the tests replicate those on the original R package.
We can make them more exhaustive.

**Comment by soleneweill** [comment]
I tried to run deseq on an AML cohort. 
I can create the deseq object but I could not get the results. I ran it with the default values and the error I have is the following. 
I can provide the data if you want to try to reproduce the error. 
Screenshot 2023-10-06 at 14 23 48

**Review by soleneweill — APPROVED** [review]
I tested  the package to run differential expression on virtual bulk for single cell data and it was working well. 
No other comments 👏

**Pull request merged**